File: leetcode/python/medium/Problem200.py
# Questions to ask:
# 1. What is the time complexity? O(m*n)
# 2. What is the space complexity? O(m*n)
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        m, n = len(grid), len(grid[0])

        def dfs(i: int, j: int):
            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != "1":
                return

            grid[i][j] = "#"

            for di, dj in [(1,0), (-1,0), (0,1), (0,-1)]:
                dfs(i+di, j+dj)

        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    dfs(i, j)
                    ans += 1
        return ans


# Problem 200
# Link: https://leetcode.com/problems/number-of-islands/description/
# Tips:
if __name__ == '__main__':
    s = Solution()

File: leetcode/golang/medium/Problem200.go
package main

func numIslands(grid [][]byte) int {
	r, c := len(grid), len(grid[0])

	directions := [][]int{[]int{1, 0}, []int{0, 1}, []int{0, -1}, []int{-1, 0}}
	var dfs func(dr, dc int)
	dfs = func(dr, dc int) {
		if 0 <= dr && dr < r && 0 <= dc && dc < c && grid[dr][dc] == '1' {
			grid[dr][dc] = '#'
			for _, d := range directions {
				dfs(dr+d[0], dc+d[1])
			}
		} else {
			return
		}
	}

	ans := 0
	for i := range r {
		for j := range c {
			if grid[i][j] == '1' {
				dfs(i, j)
				ans += 1
			}
		}
	}

	return ans
}
